Bug/64292 disabling an incomplete phase leads to inconsistencies

dombesz opened the pull request:
# Ticket
https://community.openproject.org/wp/64292

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

dombesz committed:
Always show clear button if the field is enabled.

This change is required because we do not refresh the form on every
keystroke to dynamically display the clear button, so it's better to
display it all the time as we do on the work package datepicker.

dombesz committed:
[#64292] Disabling an incomplete phase leads to inconsistencies

https://community.openproject.org/work_packages/64292

Reschedule following phases to the closest active preceding phase when activating/deactivating a phase.

This is an important step, because otherwise unset phases' default dates
will be left behind during rescheduling.

Move default start date on phases during rescheduling on phases where no finish date is set.

dombesz committed:
Move scheduling changes to the RescheduleService

dombesz committed:
Fix rubocop errors

toy commented on app/services/project_life_cycle_steps/activation_service.rb:
It is probably better to use `RescheduleService` instead of `UpdateService` after `upsert` in `persist`, unless it leads to much code duplication

dombesz replied:
I agree, changed it to use the `RescheduleService`.

toy commented on app/services/project_life_cycle_steps/update_service.rb:
```diff
@@ -48,7 +48,7 @@ def reschedule_following_phases
 
     def initial_reschedule_date
-      model.active? ? model.finish_date + 1 : model.start_date
+      model.active? ? Day.next_working(from: model.finish_date).date : model.start_date
```
I think it is better to handle finding next working day in `RescheduleService`

Also looking at `Day.next_working` I have a feeling that it is easy for it to return `nil` - it is not rare to have even few weeks of non working days on organisation level (like [CERN](https://home.cern/official-holidays#:~:text=New%20Year%E2%80%99s%20Eve-,Annual%20closure,-of%20the%20site))

dombesz replied:
~~Unfortunately it's not trivial to unify all the `initial_reschedule_date` methods in the `RescheduleService`, because the [apply_working_days_change_job.rb#L37-L40](https://github.com/opf/openproject/blob/d4e94ae8d00fbd79788fe74d3e3da1429ed94271/app/workers/projects/phases/apply_working_days_change_job.rb#L37-L40) uses a different logic which does not require finding the next working day, so moving the initial schedule date is not really a feasible option.~~

As per our sync discussion, I updated the scheduler to ensure a working day is used for setting the start date.

Your concern about non-working days longer than a week is a valid point. However it is not directly related to this PR and I would rather suggest logging a bug ticket for it to be addressed separately.

toy replied:
I didn't suggest to change `Day.next_working`, just that it may lead to failure in this case

toy commented on app/services/project_life_cycle_steps/reschedule_service.rb:
```diff
@@ -48,16 +48,24 @@ def persist(service_call)
-        date_range = calculate_date_range(from, duration: phase.duration)
-        next unless date_range
+        if phase.date_range_set?
+          next_start_date = set_phase_dates_and_next_start(phase, from)
+          from = next_start_date unless next_start_date.nil?
```
```suggestion
          from = set_phase_dates_and_next_start(phase, from) || from
```

dombesz replied:
I believe it reads a bit better to have the condition instead of having just the `||`.

toy commented on spec/services/project_life_cycle_steps/update_service_spec.rb:
```diff
@@ -165,6 +165,17 @@
 
+    context "for following phase with a start date only" do
+      let!(:phase) { create(:project_phase, project:, start_date: date, finish_date: date) }
+      let!(:following) { create(:project_phase, project:, start_date: date + 1, finish_date: nil, duration: 3) }
```
duration should be `nil` here, probably easier implicitly:
```suggestion
      let!(:following) { create(:project_phase, project:, start_date: date + 1, finish_date: nil) }
```

dombesz replied:
Having an unrelated duration makes sure we do not accidentally clear or re-calculate the duration while rescheduling with a start date only.

toy commented on spec/services/project_life_cycle_steps/update_service_spec.rb:
```diff
@@ -165,6 +165,17 @@
+      it "reschedules by moving the start date only" do
+        expect(service.call(start_date: date, finish_date: date + 1)).to be_success
+
+        expect(following).to have_attributes(start_date: date + 2, finish_date: nil, duration: 3)
```
```suggestion
        expect(following).to have_attributes(start_date: date + 2, finish_date: nil, duration: nil)
```

dombesz replied:
https://github.com/opf/openproject/pull/19058#discussion_r2132640532

toy reviewed (commented):
Looks good, what do you think about first two points?

dombesz committed:
Move following_phases to the Phase model.

dombesz committed:
Use RescheduleService in the ActivationService

dombesz committed:
Solve rubocop issues.

dombesz committed:
Align ActivationService reschedule logic to the UpdateService.

dombesz committed:
Ensure the start date is a working day when rescheduling following phases with a start date only.

Additionally remove the Day.next_working method calls when calculating
the `from` starting dates passed to the scheduler. They can be removed
because now the scheduler ensures the passed day is a working day.

toy commented on app/models/project/phase.rb:
```diff
@@ -85,25 +85,30 @@ def calculate_duration
+  def previous_phases
+    return @previous_phases if defined?(@previous_phases)
+
+    @previous_phases = project.available_phases.select { it.position < position }
```
`defined?` not needed, as it will always be an Array

```suggestion
    @previous_phases ||= project.available_phases.select { it.position < position }
```

toy commented on app/models/project/phase.rb:
```diff
@@ -85,25 +85,30 @@ def calculate_duration
-     .available_phases
-     .select { it.position < position }
-     .last
+    @following_phases = project.available_phases.select { it.position > position }
```
Same here (UI didn't allow me to select all lines)
```suggestion
    @following_phases ||= project.available_phases.select { it.position > position }
```

toy commented on app/services/project_phases/activation_service.rb:
```diff
@@ -69,24 +69,31 @@ def upsert(active:)
-      else
-        phase
-      end
+      preceding_phase || (phase.date_range_set? && phase)
```
Looks a bit hacky, probably two guards would be clearer

toy commented on spec/services/project_phases/reschedule_service_spec.rb:
```diff
@@ -146,18 +146,20 @@
+        expect(phases[1]).to have_attributes(start_date: from, finish_date: nil, duration: 5)
+        expect(phases[2]).to have_attributes(start_date: from, finish_date: nil, duration: 5)
+        expect(phases[3]).to have_attributes(start_date: nil, finish_date: nil, duration: 5)
+        expect(phases[4]).to have_attributes(start_date: from, finish_date: from + 13, duration: 10)
```
I suggest modifying the test to ensure that phases with start date will not be updated with non working day, maybe adding such phase at the end, and/or using `from` that points to non working day

toy reviewed (approved):
Looks good, just little nitpicks in production code and I suggest a bit extending the test to ensure that phase with only start date will get only working day assigned

dombesz committed:
Simplify code, add specs when a start date is rescheduled skipping non working days.
